In [38]:
from Neuron import Neuron
from NeuronLayer import NeuronLayer
from NeuronNetwork import NeuronNetwork

from sklearn.datasets import load_iris
from sklearn import datasets

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

import numpy as np
import pandas as pd
import statistics

#### Load datasets and convert to df

In [39]:
iris = datasets.load_iris()
digits = datasets.load_digits()

In [40]:
data1 = pd.DataFrame(data= np.c_[iris['data'], iris['target']],
                     columns= iris['feature_names'] + ['target'])
data2 = pd.DataFrame(data= np.c_[digits['data'], digits['target']],
                     columns= digits['feature_names'] + ['target'])

In [41]:
data1.sample()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
69                5.6               2.5                3.9               1.1   

    target  
69     1.0

In [42]:
data2.iloc[:, :-1]

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

In [43]:
data_digits = data2.iloc[:, :-1]
target_digits = data2.iloc[:, -1]

#### Classifying the flowers

In [44]:
setosa = 0
versicolor = 1
virginica = 2

#### Filter the flowers into two datasets.

flowers1 = `Setosa and Versicolour` <br>
flowers2 = `Versicolour and Verginica`


In [45]:
flowers1 = data1[data1.target != 2]

In [46]:
flowers2 = data1[data1.target != 0]

In [47]:
data_flowers1 = flowers1.iloc[:, :-1]
data_flowers2 = flowers2.iloc[:, :-1]

In [48]:
target_flowers1 = flowers1.iloc[:, -1]
target_flowers2 = flowers2.iloc[:, -1]

target_flowers1.replace(1,0, inplace=True) # resetting target to normal values. [2, 1] -> [1, 0]
target_flowers2.replace(2,1, inplace=True)

#### Manually creating the datasets for logic gates

In [49]:
AND_data = {'data':[[0, 0],
                    [0, 1],
                    [1, 0],
                    [1, 1]], 'target':[0, 0, 0, 1]}

XOR_data = {'data':[[0, 0],
                    [0, 1],
                    [1, 0],
                    [1, 1]], 'target':[0, 1, 1, 0]}

#### Split flower and digit data in train/test sets

In [50]:
# Flowers Setosa and Versicolour
flower_data1 = data_flowers1.values.tolist() # Convert dataframe to list
flower_target1 = target_flowers1.values.tolist()

X_train1, X_test1, Y_train1, Y_test1 = train_test_split(
    flower_data1, flower_target1, test_size = 0.20, random_state=1)

# Flowers Versicolour and Verginica
flower_data2 = data_flowers2.values.tolist()
flower_target2 = target_flowers2.values.tolist()

X_train2, X_test2, Y_train2, Y_test2 = train_test_split(
    flower_data2, flower_target2, test_size = 0.20, random_state=1)

# Digits
digit_data = data_digits.values.tolist()
digit_target = target_digits.values.tolist()

X_train3, X_test3, Y_train3, Y_test3 = train_test_split(
    digit_data, digit_target, test_size = 0.20, random_state=1)

### AND-gate

In [21]:
neuron_network1 = NeuronNetwork()
neuron_network1.create_layers(2,[], 1)

neuron_network1.train(AND_data['data'], AND_data['target'], l_rate=0.5, epochs=10000)

Total cost under 0.01. Ending training early at epoch: 306


In [23]:
y_pred = []
for i in AND_data['data']:
    y_pred.append(neuron_network1.feed_forward(i, perceptron=True))
accuracy = accuracy_score(AND_data['target'], y_pred)
accuracy

1.0

`100%` accuracy.

### XOR-gate

In [139]:
neuron_network3 = NeuronNetwork()
neuron_network3.create_layers(2,[2], 1)

neuron_network3.train(XOR_data['data'], XOR_data['target'], l_rate=0.5, epochs=10000)  

Total cost under 0.01. Ending training early at epoch: 1645


In [26]:
y_pred = []
for i in XOR_data['data']:
    y_pred.append(neuron_network3.feed_forward(i, perceptron=True))

accuracy = accuracy_score(XOR_data['target'], y_pred)
accuracy

1.0

**Conclusion**: Unlike a perceptron, a neuron **can** be trained on a XOR-gate as we're able to create multiple layers. <br>The  accuracy is `100%`

However there are cases where the network trains for more than 10k, 25k and even 50k epochs but is stuck at a cost higher than 0.5. <br>Using different learning rates doesn't seem to solve this problem. The easiest, most immediate solution is to increase the amount of neurons in the hidden layers from 2 to 3. This guarantees the success rate to 100%.

### Flower dataset1 (Setosa en Versicolour)

In [27]:
neuron_network4 = NeuronNetwork()
neuron_network4.create_layers(4,[], 1)
neuron_network4.train(X_train1, Y_train1, l_rate=0.5, epochs=10000) 

Total cost under 0.01. Ending training early at epoch: 1


In [28]:
y_pred = []
for i in X_test1:
    y_pred.append(neuron_network4.feed_forward(i, perceptron=True))

accuracy = accuracy_score(Y_test1, y_pred)
accuracy

1.0

**Conclusion**: Just like with the XOR-gate, a neuron **can** be trained on the first flower dataset. <br>The  accuracy is `100%`

### Flower dataset2 (Versicolour en Verginica)

In [29]:
neuron_network5 = NeuronNetwork()
neuron_network5.create_layers(4,[], 1)
neuron_network5.train(X_train2, Y_train2, l_rate=0.5, epochs=10000) 

Total cost under 0.01. Ending training early at epoch: 1


In [30]:
y_pred = []

for i in X_test2:
    y_pred.append(neuron_network5.feed_forward(i, perceptron=True))

accuracy = accuracy_score(Y_test2, y_pred)
accuracy    

1.0

**Conclusion**: Just like with the XOR-gate, a neuron **can** be trained on the second flower dataset. <br>The  accuracy is `100%`

# Digits dataset

Dataprep the targets. For example:
<br>
Target [8] turns into [0, 0, 0, 0, 0, 0, 0, 0, 1, 0]
<br>
Target [9] turns into [0, 0, 0, 0, 0, 0, 0, 0, 0, 1]

In [52]:
for counter, i in enumerate(Y_train3):
    new_target = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    index = int(i)
    new_target[index] = 1
    Y_train3[counter] = new_target 
    
for counter, i in enumerate(Y_test3):
    new_target = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    index = int(i)
    new_target[index] = 1
    Y_test3[counter] = new_target    

## Testing network to find optimal parameters
#### Testing learning rates

In [104]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.01, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.43910617229804283
CPU times: total: 5.23 s
Wall time: 5.23 s


In [108]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.3215704154299606
CPU times: total: 5.17 s
Wall time: 5.17 s


In [109]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.1, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.39067497162497394
CPU times: total: 5.3 s
Wall time: 5.29 s


In [113]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.5, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.45654931548874067
CPU times: total: 5.16 s
Wall time: 5.18 s


Learning rate `0.05` seems the most optimal of the 4

### Testing neuron amount for 1st hidden layer

In [110]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[32], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.318854425907881
CPU times: total: 10 s
Wall time: 10 s


In [111]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.24981427839919204
CPU times: total: 19.8 s
Wall time: 19.8 s


In [112]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[128], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.29221254323467755
CPU times: total: 38.4 s
Wall time: 38.4 s


An initial hidden layer with `64` neurons seems to be the most optimal as the first hidden layer

### Testing neuron amount for 2nd hidden layer

In [114]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.3788958135734751
CPU times: total: 20.1 s
Wall time: 20.1 s


In [115]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 32], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.3389125348912739
CPU times: total: 24 s
Wall time: 24.1 s


In [118]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.2877682272677743
CPU times: total: 32.1 s
Wall time: 32.2 s


In [119]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.2677023496705592
CPU times: total: 49.6 s
Wall time: 49.7 s


Since both hidden layers [64, 64] and [64, 128] are about the same, we'll go with the more efficient hidden layer. `[64, 64]`

### Testing neuron amount for 3rd hidden layer

In [120]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64, 16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.4236234608817822
CPU times: total: 34 s
Wall time: 34.2 s


In [121]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64, 32], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.3897088755145207
CPU times: total: 36.7 s
Wall time: 36.7 s


In [122]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64, 64], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.31580776492469914
CPU times: total: 43.5 s
Wall time: 43.6 s


In [123]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64, 128], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.2671146887995671
CPU times: total: 59.2 s
Wall time: 59.2 s


In [124]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64, 8], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.43226670559388264
CPU times: total: 29.9 s
Wall time: 30 s


In [125]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.42525057308943576
CPU times: total: 45.8 s
Wall time: 45.9 s


In [126]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 32], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.3958083197895224
CPU times: total: 52.3 s
Wall time: 52.3 s


In [127]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 64], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.28825409032402016
CPU times: total: 1min 4s
Wall time: 1min 4s


In [128]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 128], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.2563250833050042
CPU times: total: 1min 30s
Wall time: 1min 31s


After testing multiple networks with 3 hidden layers containing different neuron amounts, the best result seems to be the hidden layer with these neurons `[64, 64, 128]`, which gives a final cost of `0.267` after 10 epochs. Not that great, knowing that we already reach this result with hidden layers [64, 64].

### Testing neuron amount for 4th hidden layer [FINAL]

In [129]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 128, 16], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.42341093678340386
CPU times: total: 1min 29s
Wall time: 1min 30s


In [130]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 128, 32], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.37582565036987353
CPU times: total: 1min 36s
Wall time: 1min 36s


In [131]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 128, 64], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.3913661431778733
CPU times: total: 1min 48s
Wall time: 1min 49s


In [132]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 128, 128, 128], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=10) 

Training finished after: 10 epochs
Total cost: 0.36093652671530335
CPU times: total: 2min 14s
Wall time: 2min 14s


The best cost we get after 10 epochs with 4 hidden layers is around `0.36`, which is worse than we previously got.
<br>
Therefore we will keep the best result. `[64, 64]`.

### Testing the network

In [133]:
neuron_network = NeuronNetwork()
neuron_network.create_layers(64,[64, 64], 10)
%time neuron_network.train(X_train3, Y_train3, l_rate=0.05, epochs=100) 

Training finished after: 100 epochs
Total cost: 0.14531091848972214
CPU times: total: 5min 34s
Wall time: 5min 35s


In [134]:
y_pred = []

for i in X_test3:
    output = neuron_network.feed_forward(i)
    lst = [0] * len(output)
    max_index = output.index(max(output))
    lst[max_index] = 1
    
    y_pred.append(lst)

accuracy = accuracy_score(Y_test3, y_pred)
accuracy    

0.8138888888888889

After 100 epochs we get an accuracy of 0.81 = `81%`. 
<br>
Not too bad, but what if we up it to 1000 epochs? Let's find out.

In [135]:
neuron_network01 = NeuronNetwork()
neuron_network01.create_layers(64,[64, 64], 10)
%time neuron_network01.train(X_train3, Y_train3, l_rate=0.05, epochs=1000) 

Training finished after: 1000 epochs
Total cost: 0.1860391479705559
CPU times: total: 57min 23s
Wall time: 57min 44s


In [136]:
y_pred = []

for i in X_test3:
    output = neuron_network01.feed_forward(i)
    lst = [0] * len(output)
    max_index = output.index(max(output))
    lst[max_index] = 1
    
    y_pred.append(lst)

accuracy = accuracy_score(Y_test3, y_pred)
accuracy    

0.7694444444444445

After 1000 epochs and almost an hour, we get an accuracy of 0.77 = `77%`. 
<br>
Less than what we're able to reach with just 100 epochs.

**Conclusion**: a neural network containing 2 hidden layers with 64 neurons each gives the best result.
<br>
And a learning rate of 0.05 seems to be the most optimal. 
<br>
After running 100 epochs, an accuracy of `~80%` can be expected. Running more epochs doesn't seem to be beneficial.